Update EIP-5792: Propose changes to EIP-5792 (#8826)
ethereum/EIPs: EIPS/eip-5792.md

The items in the `calls` field are simple `{to, data, value}` tuples.

The capabilities field is how an app can communicate with a wallet about capabilities a wallet supports. For example, this is where an app can specify a paymaster service URL from which an [ERC-4337](./eip-4337.md) wallet can request a `paymasterAndData` input for a user operation.

Each capability defined in the "capabilities" member can define global or call specific fields.
These fields are set inside this capability's entry in the `capabilities` object.
Each entity in the `calls` field may contain an optional `capabilities` object.
This object allows the applications to attach a capability-specific metadata to individual calls.

Unless these requirements are explicitly overriden by a certain `capability`, the wallet must adhere to the following rules.
Note that such a `capability` is not in the scope of this EIP and should be defined in its own ERC.

The wallet:

* MUST send the calls in the order specified in the request
* MUST send the calls on the same chain identified by the request's `chainId` property
* MUST send the calls from the address specified in the request's `from` property, if provided

>   * If `from` is not provided the wallet SHOULD present the user with an opportunity to view and change the `from` address during confirmation

Reviewer: This is a little hard to parse-- if it's not provided there's nothing to change? do you mean:
```suggestion
  * If `from` is not provided the wallet SHOULD present the user with an opportunity to select a `from` address during confirmation
```

Author: Replaced 'change' with 'select' to indicate this field is not user-editable

* MUST submit multiple calls as an atomic unit in a single transaction
* MUST NOT send any calls from the request if the user rejects the request
* MAY revert all calls if any call fails
* MUST not execute any further calls after a failed call
* MAY reject the request if the from address does not match the enabled account
* MAY reject the request if one or more calls in the batch is expected to fail, when simulated sequentially
* MUST reject the request if it contains a `capability` (either top-level or call-level) that is not supported by the wallet and the `capability` is not explicitly marked as optional.
  * Applications may mark a capability as optional by setting `optional` to `true`. See the RPC Specification section below for details.

If provided, the wallet MUST respect the `id` field and return it in the response.

Identifiers, whether provided by the app or generated by the wallet, MUST be a unique string up to 4096 bytes (8194 characters including leading `0x`).

App-provided `id`s MUST be unique per sender per app, where each "app" SHOULD be identified by their domain.

Wallets MUST reject requests with duplicate `id`s.

> Within 24 hours from the corresponding `wallet_sendCalls`, wallets SHOULD return a call-batch status when `wallet_getCallsStatus` is called with the same `id`.

Reviewer 2: Does it mean that the wallet should store the `wallet_sendCalls` result for 24hours? What's the reasoning for that?

Author: But, if the wallets do not store and return the results of `wallet_sendCalls`, how would an app ever get them?

Reviewer 2: My question is about the reasoning for 24 hours. Wondering if it's just an arbitrary number or sth else.
I'd even consider changing SHOULD to MUST due to having `wallet_getCallsStatus` calls guaranteed

Author: Ok, got it. Yes, "24 hours" is an arbitrary number, and that is why I wrote SHOULD - if the value is not available after, say, 19 hours it is probably not such a big deal. I guess in the vast majority of cases the `wallet_getCallsStatus` will be only relevant for a couple of minutes. 
We can change it to MUST if we want to be strict about it, and we can increase the time window if there is a potential use-case for it.

Reviewer: it's odd grammatically, because it could be read as "wallets are expected to return a call status at some point in the next 24 hours" or it cold be read as "apps can assume they're never getting a call-batch if 24 hours pass". maybe worth breaking wallet behavior and app behavior into 2 distinct sentences?

Author: > "wallets are expected to return a call status at some point in the next 24 hours" or it cold be read as "apps can assume they're never getting a call-batch if 24 hours pass".

Frankly, I feel fine leaving any behaviour after the 24-hour window udefined, as this is an extremely long time for a transaction.
As for the meaning within the 24 window, clearly the wallet will want to return a status as soon as possible, and the dapps will not be able to work with a wallet that is not able to do so.
If you can suggest a more clear definition of wallet/app expectations, I would appreciate if you could make a PR.

### `wallet_sendCalls`

Requests that a wallet submits a batch of calls. `from` and `chainId` are top-level properties rather than per-call properties because all calls should be sent from the same sender and on the same chain, identified by [EIP-155](./eip-155.md) integers expressed in hexadecimal notation.
The items in the `calls` field are simple `{to, data, value}` tuples.

The capabilities field is how an app can communicate with a wallet about capabilities a wallet supports. For example, this is where an app can specify a paymaster service URL from which an [ERC-4337](./eip-4337.md) wallet can request a `paymasterAndData` input for a user operation.

Each capability defined in the "capabilities" member can define global or call specific fields.

> These fields are set inside this capability's entry in the `capabilitiesData` object.

Reviewer 3: I’m curious why is this termed as `capabilitiesData` instead of just `capabilities` for consistency? Isn’t this the same thing as global capabilities but on a per-call level?

Each entity in the `calls` field may contain an optional `capabilitiesData` object.
This object allows the applications to attach a capability-specific metadata to individual calls.

The wallet:

* MUST send the calls in the order specified in the request
* MUST send the calls on the same chain identified by the request's `chainId` property
* MUST NOT await for any calls to be finalized to complete the batch
* MUST NOT send any calls from the request if the user rejects the request
* MAY revert all calls if any call fails
* MAY send all calls as part of one or more transactions, depending on wallet capability
* SHOULD stop executing the calls if any call fails
* MAY reject the request if the from address does not match the enabled account
* MAY reject the request if one or more calls in the batch is expected to fail, when simulated sequentially

> * MUST reject the request if it contains a `capability` that is not supported by the wallet

Reviewer 4: Hmm I wonder if this is best 🤔 . E.g. why not allow specifying a paymaster which the app would like to be used but is not a necessity.

Reviewer 5: How a wallet can tell if the cap (e.g. paymaster) is required by this app or optional?
The app can always query the wallet caps first, to decide how to continue.

Reviewer 6: capability specs should specify whether they're required or not / how optional ones are handled. paymaster, for example, notes that the app must not assume the provided paymaster service is the one that ends up being used.

so if an app wants the capability to be required, the capability spec should include a `required` field, and that is how a wallet can tell if the capability is required by the app. i dont think we should blanket all capabilities to be required by default.

Author: There are two issues raised here:

1. App providing a "capability" that is *not supported* by the wallet:

There is a `wallet_getCapabilities` method that provides the application with a list of *supported* "capabilities".
Shouldn't the app that ignores that list and submits a `wallet_sendCalls` with a "capability" that is not supported be considered to behave incorrectly? I think it should reject the call, which would prevent dapps from yoloing their `sendCalls` requests.

2. Wallet discretion to use or ignore a *supported* "capability" requested by the dapp:

As the API defined here deals with real value and assets, making capabilities `required` by default makes a lot of sense to me. If using a paymaster for a transaction is indeed optional, it is trivial to specify `optional: true`, but forgetting to set `required: true` for a security-oriented capability may be catastrophic. 

What do you think?

Reviewer 4: So @forshtat is the conclusion here that each capability *could* have an `optional: ` flag? But passing a capability the wallet doesn't have would be hard failure?

Author: > So @forshtat is the conclusion here that each capability _could_ have an `optional: ` flag? But passing a capability the wallet doesn't have would be hard failure?

In my personal opinion, yes, this seems pretty reasonable.

#### `wallet_sendCalls` RPC Specification

```typescript
type SendCallsParams = {
  version: string;
  from: `0x${string}`;
  chainId: `0x${string}` | undefined; // Hex chain id
  calls: {
    to?: `0x${string}` | undefined;
    data?: `0x${string}` | undefined;
    value?: `0x${string}` | undefined; // Hex value

>     capabilitiesData?: Record<string, any> | undefined;

Reviewer 4: hmm, I don't love how this kind of muddles the simplicity of the calls struct. Is the complaint that capabilities makes it hard to specify info on a per call basis. Could we get a better example of when this would be needed 🤔 

> One example could be an "optional flag" capability that would allow a Smart Contract Wallet to proceed with the batch in case of a revert

This should be possible with a general capability? Or you're saying one call in particular can revert?

Reviewer 5: You don't have to implement it, if our wallet doesn't have any per-call capability. 
Also, an app doesn't have to know about it, unless it needs that capability.
The idea that if someone ever want to add per-call capability, it doesn't mess directly with the "call" structure, or with such params of another capability.

Author: > Is the complaint that capabilities makes it hard to specify info on a per call basis?

Yes, that pretty much sums it up. Of course, the `capability` could provide an array of details, which would have to match the `calls` array's length exactly and contain `null` values for calls that do not require extra parameters, which is functionally equivalent but seems like a much more quirky API to me.

> Or you're saying one call in particular can revert?

In my example, yes, I was talking about an "atomic" batch with a single "optional" call that should not revert the batch;
It is hard to reason about potential future additional examples, so these are a little hypothetical, but in my opinion, worth considering:
If a "per-call gas limit capability" is defined, it can be used to provide a per-call gas limit;
In case a "multi-chain capability" is defined, the 'capabilitiesData' field can be used to specify a per-call 'chainId' and L2-specific parameters, something like `gas_per_pubdata_limit`;
In case a "conditional call capability" is defined, it can provide some description of a condition to include or exclude a call from the batch;
etc.

Does this make sense? Or should dapps never need to go to the trouble of defining per-call parameters for a `sendCalls` request?

The items in the `calls` field are simple `{to, data, value}` tuples.

The capabilities field is how an app can communicate with a wallet about capabilities a wallet supports. For example, this is where an app can specify a paymaster service URL from which an [ERC-4337](./eip-4337.md) wallet can request a `paymasterAndData` input for a user operation.

Each capability defined in the "capabilities" member can define global or call specific fields.
These fields are set inside this capability's entry in the `capabilities` object.
Each entity in the `calls` field may contain an optional `capabilities` object.
This object allows the applications to attach a capability-specific metadata to individual calls.

The wallet:

* MUST send the calls in the order specified in the request
* MUST send the calls on the same chain identified by the request's `chainId` property
* MUST NOT await for any calls to be finalized to complete the batch

> * MUST submit multiple calls as part of a single transaction, unless instructions for sequential execution are provided via a `capability`.

Author: Please see:
https://github.com/ethereum/EIPs/pull/8826/commits/0de3f96d90cf95929bb85b193eb62bb7f0bff33a

* MAY revert all calls if any call fails
* MAY send all calls as part of one or more transactions, depending on wallet capability
* SHOULD stop executing the calls if any call fails
* MAY reject the request if the from address does not match the enabled account
* MAY reject the request if one or more calls in the batch is expected to fail, when simulated sequentially
* MUST reject the request if it contains a `capability` (either top-level or call-level) that is not supported by the wallet and the `capability` is not explicitly marked as optional.
  * Capabilities (both top-level and call-level) MUST be marked as optional by applications sending requests to wallets by passing an `optional` parameter set to `true` for each `capability` that is not required.

#### `wallet_sendCalls` RPC Specification

```typescript
type SendCallsParams = {
  version: string;
  from: `0x${string}`;
  chainId: `0x${string}` | undefined; // Hex chain id
  calls: {
    to?: `0x${string}` | undefined;
    data?: `0x${string}` | undefined;
    value?: `0x${string}` | undefined; // Hex value
    capabilities?: Record<string, any> | undefined;
  }[];

>   capabilities?: Record<string, any> | undefined;

Author: https://github.com/ethereum/EIPs/pull/8826/commits/534c81b454d1906edce159ee99b4aed4d0f463f6

  {
    "version": "1.0",
    "from": "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
    "chainId": "0x01",
    "calls": [
      {
        "to": "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
        "value": "0x9184e72a",
        "data": "0xd46e8dd67c5d32be8d46e8dd67c5d32be8058bb8eb970870f072445675058bb8eb970870f072445675"
      },
      {
        "to": "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
        "value": "0x182183",
        "data": "0xfbadbaf01"
      }
    ],
    "capabilities": {
      "paymasterService": {
        "url": "https://...",
        "optional": true
      }
    }
  }
]
```

Note that since the `paymasterService` `capability` is marked as optional, wallets that do not support it can still process and handle the request to the best of their ability. If this `optional` field were set to `false` or absent from the request, wallets that do not support the `capability` MUST reject the request.

##### `wallet_sendCalls` Example Return Value

> The identifier MUST be unique 64 bytes represented as a hex encoded string.

Author: I think a 64 bit identifier can be assumed globally unique for the purpose of this EIP, but we can add "for the current session".
But, as you've mentioned, we did not yet define what a "session" actually is for the `wallet_sendCalls`.

> The identifier MUST be unique 64 bytes represented as a hex encoded string.

Author: We could also make it explicitly more limited, something like "The identifier MUST be unique for a given RPC provider within a 24 hour window. It is 64 bytes long and is represented as a hex encoded string."

What do you think? The 24 hour window comes from this commit:

https://github.com/ethereum/EIPs/commit/77f32cc65c2bf6c2827d25ac19bef07ed046ec52

> The identifier MUST be unique 64 bytes represented as a hex encoded string.

Reviewer 3: What is the rationale behind a strict 64-byte identifier? 🤔

> For a given session, users should be able to call `wallet_getCallsStatus` with this value and expect a call-batch status to be returned.

Author: Well, there is no strong definition for a "session" in this case. Is a 24-hour window a better suggestion? The rule is just a "SHOULD" so it is not binding anyways.

https://github.com/ethereum/EIPs/pull/8826/commits/77f32cc65c2bf6c2827d25ac19bef07ed046ec52

The `capabilities` object allows the wallets to attach a capability-specific metadata to the response.

```json
{
  "id": "0x00000000000000000000000000000000000000000000000000000000000000000e670ec64341771606e55d6b4ca35a1a6b75ee3d5145a99d05921026d1527331",
}
```

### `wallet_getCallsStatus`

Returns the status of a call batch that was sent via `wallet_sendCalls`.
The identifier of the batch is the value returned from the `wallet_sendCalls` RPC.
Note that the `receipts` objects of this method's response is a strict subset of the object returned by `eth_getTransactionReceipt`.

The `capabilities` object allows the wallets to attach a capability-specific metadata to the response.

* The receipts in the `receipts` field MUST be in order they are included onchain.
* The `receipts` field MUST contain receipts for all calls in a batch that were included onchain, including reverted ones.

> * If a wallet executes multiple calls atomically (i.e. in a single transaction), `wallet_getCallsStatus` MUST return a single receipt, corresponding to the transaction in which the calls were included.

Reviewer 5: IIUC:
- if the sendCalls performs an atomic transaction (as is the default for this ERC), then there is a single receipt for the transaction itself - not for each inner call.
- if there are separate transactions for the calls (which is not fully specified in this ERC) - then there is a receipt for each call.
> multisig wallet, where each signature is recorded in a transaction, then applied atomically once the quorum is reached.
If each multisig owner separately submits a transaction with its own signature, then each of them is a separate transaction, originated by separate wallets. The last owner signature is batched transaction with the multisig execute that uses all these partial signatures.

> * If a wallet executes multiple calls atomically (i.e. in a single transaction), `wallet_getCallsStatus` MUST return a single receipt, corresponding to the transaction in which the calls were included.

Author: Rephrased a little:
https://github.com/ethereum/EIPs/pull/8826/commits/a3d61d369dce4e9fe1fa3847942f23a6e9fd2135

  {
    "version": "1.0",
    "from": "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
    "chainId": "0x01",
    "calls": [
      {
        "to": "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
        "value": "0x9184e72a",
        "data": "0xd46e8dd67c5d32be8d46e8dd67c5d32be8058bb8eb970870f072445675058bb8eb970870f072445675"
      },
      {
        "to": "0xd46e8dd67c5d32be8058bb8eb970870f07244567",
        "value": "0x182183",
        "data": "0xfbadbaf01"
      }
    ],
    "capabilities": {
      "paymasterService": {
        "url": "https://...",
        "optional": true
      }
    }
  }
]
```

Note that since the `paymasterService` `capability` is marked as optional, wallets that do not support it will still process and handle the request as if the capability had not been present. If this `optional` field were set to `false` or absent from the request, wallets that do not support the `capability` MUST reject the request.

##### `wallet_sendCalls` Example Return Value

The identifier MUST be unique 64 bytes represented as a hex encoded string.

> Within 24 hours from the corresponding `wallet_sendCalls`, wallets SHOULD return a call-batch status when `wallet_getCallsStatus` is called with this value.

Reviewer 7: This line should be moved from the example subsection to the spec part.

type GetCallsParams = string;

type GetCallsResult = {
  batchId: `0x${string}`;
  chainId?: `0x${string}`;
  batchStatus: string; // See "Status Codes"
  receipts?: {
    logs: {
      address: `0x${string}`;
      data: `0x${string}`;
      topics: `0x${string}`[];
    }[];
    status: `0x${string}`; // Hex 1 or 0 for success or failure, respectively
    blockHash: `0x${string}`;
    blockNumber: `0x${string}`;
    gasUsed: `0x${string}`;
    transactionHash: `0x${string}`;
  }[];

>   capabilitiesData?: Record<string, any> | undefined;

Reviewer 4: I find it kinda confusing that this is the same phrase as on the request, but would likely be different data?

Author: I fully support renaming this parameter. `capabilitiesResultsData` is a bit long, how about just `capabilitiesResults` or `capabilitiesStatus`?

Reviewer 4: I prefer clarity even if verbose so I think I like `capabilitiesResultsData`

Author: > I prefer clarity even if verbose so I think I like `capabilitiesResultsData`

Done.

};
```

> ##### Status Codes for `batchStatus` field

Reviewer 8: Maybe we should define what a batch is first? it's the first time we use it

The purpose of the `batchStatus` field is to provide a short summary of the current status of the batch.
It provides some off-chain context to the array of inner transaction `receipts`.

> | Name      | Description                                                                                                          |

Reviewer 4: I like this idea, thoughts @jxom @lukasrosario ?

Reviewer 6: yeah i think this makes sense if we're talking about single chain. would be tough if multichain.

also not clear to me where `DISCARDED` would come up. @forshtat can you share more on how you were thinking this might come up?

Author: Yes, it sounds similar to "FAILURE" but the difference I had in mind was that a "FAILURE" means the action failed due to on-chain reverts, and "DISCARDED" means the wallet gave up on trying to send the batch to the network.
Maybe the names do not make the distinction clear enough, or do you think the distinction is not worth the extra status code?

One example could be a batch that specifies a `paymasterService` capability, and given the wallet need to go through these steps before sending anything on-chain:
<img width="756" alt="Screenshot 2024-08-28 at 18 16 29" src="https://github.com/user-attachments/assets/a0255df8-fa3f-47a8-a08d-0cb768faf217">

The wallet cannot block the `wallet_sendCalls()` for all this time, so returns some `identifier` and proceeds with preparing the UserOperation. 
During the preparation phase the `wallet_getCallsStatus` returns PENDING as nothing really happened.
However, if the "Paymaster Service" fails with an error at some stage, wallet could keep retrying to reach it for some time, and the transaction has not technically "failed", but after 2 minutes of trying to reach the "Paymaster Service" it gives up.

Another example is a UserOperation that has some condition in its validation function. Canonical ERC-4337 limits this use-case a lot, but does not completely prevent it. A wallet may want to accept a batch that is not yet valid, check if it has become valid during the next 2 hours, and mark the entire operation as "DISCARDED" after that.

Does this makes sense?

|-----------|----------------------------------------------------------------------------------------------------------------------|
| PENDING   | Batch has been received by the wallet but has not completed execution on-chain                                       |
| SUCCESS   | Batch has been included on-chain without reverts, receipts array contains info of all calls                          |

> | PARTIAL   | Batch has been included on-chain only partially, the rest is either reverted or invalid                              |

Reviewer 8: ```suggestion
| PARTIAL   | Batch has been included on-chain only partially, the rest is either reverted or invalid. Receipts array contains info of all successful calls             |
```

type GetCallsParams = string;

type GetCallsResult = {

>   batchId: `0x${string}`;

Reviewer 6: callsId

Reviewer 6: should keep the name the same

Author: Done.

Reviewer 3: Super nit: kinda think `id` is even sufficient enough and maybe more consistent (ie. we don’t call it `transactionHash` on the transaction schema)

  chainId?: `0x${string}`;
  batchStatus: string; // See "Status Codes"
  receipts?: {
    logs: {
      address: `0x${string}`;
      data: `0x${string}`;
      topics: `0x${string}`[];
    }[];
    status: `0x${string}`; // Hex 1 or 0 for success or failure, respectively
    blockHash: `0x${string}`;
    blockNumber: `0x${string}`;
    gasUsed: `0x${string}`;
    transactionHash: `0x${string}`;
  }[];

>   capabilitiesResultsData?: Record<string, any> | undefined;

Reviewer 3: Should this just be called `capabilities` for consistency?

};
```

##### Status Codes for `batchStatus` field

The purpose of the `batchStatus` field is to provide a short summary of the current status of the batch.
It provides some off-chain context to the array of inner transaction `receipts`.

| Name      | Description                                                                                                          |
|-----------|----------------------------------------------------------------------------------------------------------------------|

> | PENDING   | Batch has been received by the wallet but has not completed execution on-chain                                       |

Reviewer 9: How do you feel about possibly expanding `PENDING` to capture more of the offchain context around the pending state?

For example

- `PENDING`: Batch has been received by the wallet but has not completed execution on-chain 
- `UNDERPRICED`: Batch has been received by the wallet but the gas prices are too low for the current chain conditions
- `STALLED`: Batch has been received by the wallet but is stuck for non-pricing related reasons; i.e. ERC-4337 bundler is down. The definition of 'stuck' is left up to the wallet, i.e. batch is priced appropriately but has not been included in >60s

Reviewer 6: @forshtat would be good to get your thoughts here. imo `UNDERPRICED` specifically seems like a reasonable addition, as it could be retried.

@ajhodges i'm not a huge fan of the ambiguity for `STALLED` though. to me this is effectively still `PENDING`, and each wallet should have their own interpretations of `PENDING` taking "too long". if we and others feel strongly about it being its own status, i would prefer to be more opinionated in the spec and define when a `PENDING` batch becomes `STALLED`, instead of leaving it open-ended.

Reviewer 3: What if we just left it as `PENDING` and `SUCCESS`, and had a sibling `statusCode` or `detail` property that could be extensible enough to accommodate for what ever addition? 
This would make it a lot easier to coerce on frontend apps.

Author: > `UNDERPRICED` specifically seems like a reasonable addition, as it could be retried.

I don't have any strong feelings against `UNDERPRICED` or any other status code we can clearly define.

About `UNDERPRICED` specifically, I guess its difference with the regular `PENDING` state is that the user needs to confirm a modification to the original request? And if the wallet could just increase the price internally and retry, it would not be different from `PENDING`, would it?

If so, maybe a better broader category would be `ACTION_REQUIRED` or something like that, what do you think?

Reviewer 3: Yeah, I guess `statusCode` might be more realistic than `detail` as an application can choose which message they want to display to an end-user (if anything). 

Perhaps this could be a similar pattern to HTTP status codes where `2xx` ≈ success, `4xx` ≈ client error, `5xx` ≈ server error, etc. For this case, maybe `1xx` represents `PENDING` sub-statuses, `2xx` represents `CONFIRMED` sub-statuses, and so on. For example, a `status` of `PENDING` and a `statusCode` of `101` could mean “stalled”, and an application could either choose to show a generic loading indicator, or a loading indicator with “this tx is taking a while, pls wait”. 

I guess the motivation of a sub-status field like `statusCode` is that it would make the branching logic within an application so much easier since 5792 is an ERC for applications.

Reviewer 9: >I guess the motivation of a sub-status field like statusCode is that it would make the branching logic within an application so much easier — they can rely on the simplicity of status and choose to show extra information by utilizing statusCode.

Yep, exactly. As an app, I want to know if things are healthy. Is my transaction stuck? Is there anything I can do about it if it's stuck? (i.e. underpriced vs stalled). I think the display logic/messaging to the user should be left up to the app, but the wallet can oftentimes provide additional context about what happened to make the UX a little better.

I agree `UNDERPRICED` is maybe a bit vague. But the wallet should be the expert on the current chain conditions and the bundle pricing so it's best suited to detect if the calls should be retried (and we shouldn't prescribe any conditions in this EIP). So maybe `RETRYABLE` is a better sub-status than `UNDERPRICED`, signifying that the calls might go through quicker if you retry. I think I like this better than `ACTION_REQUIRED` because technically an underpriced tx can still go through if chain conditions change

type GetCallsParams = string;

type GetCallsResult = {
  id: `0x${string}`;
  chainId?: `0x${string}`;

>   status: string; // See "Status Codes"

Reviewer 6: ah true, just missed this when making the status change, will update.

  receipts?: {
    logs: {
      address: `0x${string}`;
      data: `0x${string}`;
      topics: `0x${string}`[];
    }[];
    status: `0x${string}`; // Hex 1 or 0 for success or failure, respectively
    blockHash: `0x${string}`;
    blockNumber: `0x${string}`;
    gasUsed: `0x${string}`;
    transactionHash: `0x${string}`;
  }[];
  capabilities?: Record<string, any> | undefined;
};
```

##### Status Codes for `status` field

The purpose of the `status` field is to provide a short summary of the current status of the batch.
It provides some off-chain context to the array of inner transaction `receipts`.

| Code | Description |
|------|-------------|
| 100  | Batch has been received by the wallet but has not completed execution onchain (pending) |
| 200  | Batch has been included onchain without reverts, receipts array contains info of all calls (confirmed) |
| 400  | Batch has not been included onchain and wallet will not retry (offchain failure) |

> | 500  | Batch reverted completely and only changes related to gas charge may have been included onchain (chain rules failure) |

Author: Added a 600 status code, what do you think? We can also make it, say, `501`.
https://github.com/ethereum/EIPs/pull/8826/commits/7701a6445691009d5ffd16b54b87de6796802683

type GetCallsParams = string;

type GetCallsResult = {
  version: string;
  id: `0x${string}`;
  chainId: `0x${string}`;
  status: number; // See "Status Codes"
  receipts?: {
    logs: {
      address: `0x${string}`;
      data: `0x${string}`;
      topics: `0x${string}`[];
    }[];
    status: `0x${string}`; // Hex 1 or 0 for success or failure, respectively
    blockHash: `0x${string}`;
    blockNumber: `0x${string}`;
    gasUsed: `0x${string}`;
    transactionHash: `0x${string}`;
  }[];

>   capabilities?: Record<string, any>;

Reviewer: what happens if these caps changed since last interaction? i'm curious what these are doing here.  just for matching?

Author: > L153: The `capabilities` object allows the wallets to attach a capability-specific metadata to the response.

Meaning, this field contains any data specific to a requested "capability", not the list of supported capabilites as in `wallet_getCapabilities`.

This method accepts a call bundle identifier returned by a `wallet_sendCalls` call.

```json
["0xe670ec64341771606e55d6b4ca35a1a6b75ee3d5145a99d05921026d1527331"]
```

### `wallet_getCapabilities`

This RPC allows an application to request capabilities from a wallet (e.g. batch transactions, paymaster communication), without distinct discovery and permission requests. For more on the difference between requesting capabilities and discovering features, see the ["Privacy Considerations" section](#privacy-considerations).

> This method SHOULD return an error if the user has not already authorized a connection between the application and the requested address.

Author: There is no definition of specific errors or error codes for any of the APIs, we could either define all of them explicitly or leave it to the wallets. Should we?

> This method SHOULD return an error if the user has not already authorized a connection between the application and the requested address.

Reviewer 6: for this one specifically i would suggest 4100 unauthorized. i dont think any others need to be defined ? as i believe they are covered by the other 1193 errors + getCallsStatus errors

> This method SHOULD return an error if the user has not already authorized a connection between the application and the requested address.

Author: https://github.com/ethereum/EIPs/pull/8826/commits/1be7d7d6b8b1e028ed64b27f27c3e0a48f208f04

Capabilities are returned in key/value pairs, with the key naming a capability and a value conforming to a shape defined for that name, in an object keyed to the relevant [EIP-155](./eip-155.md) `chainId` expressed in hexadecimal notation.
Capabilities are nested in per-chain objects because wallets may support different capabilities across multiple chains authorized in a given session.

> All values for the "capability" fields MUST contain Semantic Version value `version`.

Reviewer 3: Yeah, would make it simpler.

> Like the illustrative examples given above and other capabilities to be defined in future EIPs, the capability to execute calls delivered via the above-defined methods in a single transaction can be attested by the wallet in boolean form.

Author: This section has been removed now.

> Note that this capability is purely declarative. I.e. this capability is not meant to be paired with a `wallet_sendCalls` request (at either top-level nor call-level), as this EIP specifies that multiple calls MUST, by default, be executed as part of a single transaction.

Author: This section has been removed now.

> Note that this capability is purely declarative. I.e. this capability is not meant to be paired with a `wallet_sendCalls` request (at either top-level nor call-level), as this EIP specifies that multiple calls MUST, by default, be executed as part of a single transaction.

Author: This section has been removed now.

## Security Considerations

> App developers MUST treat each call in a batch as if the call were an independent transaction. In other words, there may be additional untrusted transactions between any of the calls in a batch. The calls in the batch MAY also be included in separate, non-contiguous blocks. There is no constraint over how long it will take all the calls in a batch to be included. Apps MUST encode deadlines and timeout behaviors in the smart contract calls, just as they do today for transactions, including ones otherwise bundled. App developers MUST NOT assume that all calls will be sent in a single transaction.

Reviewer 10: > MUST submit multiple calls as an atomic unit in a single transaction

This can be removed, but I'll steal it for flow-control
